dynamiqs/qarrays/sparse_qarray.py

```diff
@@ -264,6 +265,39 @@ def __rmatmul__(self, other: Array) -> Array:
 
         return NotImplemented
 
+    def _kronecker_dia(self, other: SparseQArray) -> SparseQArray:
+        N = other.diags.shape[-1]
+        out_offsets = jnp.ravel(
+            jnp.asarray(self.offsets) * N + jnp.asarray(other.offsets)[:, None],
+            order='F',
+        )
+        out_diags = jnp.kron(self.diags, other.diags)
+        return tuple(out_offsets), out_diags
+
+    def _clean_kronecker_dia(
+        self, offsets: tuple[int, ...], diags: ArrayLike
+    ) -> SparseQArray:
+        offsets = jnp.sort(jnp.asarray(offsets))
+        unique_offsets, inverse_indices = jnp.unique(offsets, return_inverse=True)
+        count = unique_offsets.shape[0]
+        out_diags = jnp.zeros((count, diags.shape[-1]))
+        for i in range(count):
+            mask = inverse_indices == i
+            out_diags = out_diags.at[i, :].set(jnp.sum(diags[mask, :], axis=0))
+        return SparseQArray(
+            tuple(o.item() for o in unique_offsets), out_diags, self.dims
+        )
+
+    def __and__(self, other: Array) -> QArray:
+        if isinstance(other, SparseQArray):
+            temp_o, temp_d = self._kronecker_dia(other=other)
+            return self._clean_kronecker_dia(temp_o, temp_d)
+
+        elif isinstance(other, Array):
```

Comment:
```suggestion
        elif isinstance(other, get_args(QArrayLike)):
```
and probably need a call to `asqarray(other)` or `jnp.asarray(other)` below too to convert other formats to a `QArray`.

Reply:
@bsaoptima See this comment too regarding the need to call `asqarray(other)`